contracts/exchange/Exchange_ERC20_1155.cairo

```diff
@@ -0,0 +1,507 @@
+    ERC1155_transfer_from,
+    ERC1155_batch_transfer_from,
+    ERC1155_mint,
+    ERC1155_burn,
+    ERC1155_set_approval_for_all,
+    ERC1155_assert_is_owner_or_approved,
+)
+
+#FIXME Non-reentrant
+
+# Contract Address of ERC20 address for this swap contract
+@storage_var
+func currency_address() -> (address : felt):
+end
+
+# Contract Address of ERC1155 address for this swap contract
+@storage_var
+func token_address() -> (address : felt):
+end
+
+# Current reserves of currency
+@storage_var
+func currency_reserves(token_id : felt) -> (reserves : Uint256):
+end
+
+# Total issued LP totals
+@storage_var
+func lp_reserves(token_id : felt) -> (total : Uint256):
+end
+
+# Liquidity pool rewards in thousandths (e.g. 15 = 1.5% fee)
+@storage_var
+func lp_fee_thousands() -> (lp_fee_thousands : Uint256):
+end
+
+# Note: We use ERC1155_balanceOf(contract_address) to record token reserves
+
+@constructor
+func constructor {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr,
+    }(
+        currency_address_: felt,
+        token_address_: felt,
+        lp_fee_thousands_: Uint256,
+    ):
+        currency_address.write(currency_address_)
+        token_address.write(token_address_)
+        lp_fee_thousands.write(lp_fee_thousands_)
+    return ()
+end
+
+#
+# Liquidity
+#
+
+@external
+func initial_liquidity {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr,
+    }(
+        currency_amount: Uint256,
+        token_id: felt,
+        token_amount: Uint256,
+    ):
+        alloc_locals
+        let (caller) = get_caller_address()
+        let (contract) = get_contract_address()
+
+        let (token_address_) = token_address.read()
+        let (currency_address_) = currency_address.read()
+
+        # Only valid for first liquidity add
+        let (currency_reserves_) = currency_reserves.read(token_id)
+        assert currency_reserves_ = Uint256(0, 0)
+
+        IERC20.transferFrom(currency_address_, caller, contract, currency_amount)
+        tempvar syscall_ptr :felt* = syscall_ptr
+        IERC1155.safeTransferFrom(token_address_, caller, contract, token_id, token_amount.low)
+
+        # Assert otherwise rounding error could end up being significant on second deposit
+        let (ok) = uint256_le(Uint256(1000, 0), currency_amount) #FIXME
+        assert_not_zero(ok)
+
+        # Update currency  reserve size for Token id before transfer
+        currency_reserves.write(token_id, currency_amount)
+
+        # Initial liquidity is amount deposited
+        lp_reserves.write(token_id, currency_amount)
+
+        # Mint LP tokens
+        ERC1155_mint(caller, token_id, currency_amount.low)
+
+        #TODO emit LP Added Event
+
+    return ()
+end
+
+
+@external
+func add_liquidity {
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr,
+    }(
+        max_currency_amount: Uint256,
```

Comment:
What is this argument for? Since this is a publicly callable function, one could just pass max Uint256, no? What am I missing?

Reply:
This is the maximum amount of currency (ERC20) the caller wants to add to the LP. 
When adding to the LP, the user must supply currency and token amounts in the same proportion that is currently in the pool. e.g. LP changes shouldn't affect the `x * y = k` calc. 
As the price may fluctuate between sending and execution (if other people are buying and selling at the same time), the user specifies a maximum amount of currency they are willing to add to the LP and a fixed amount of token. The contract determines the correct LP price for the currency and uses only that much, refunding the rest.